### ISSUE TITLE
Performance issue in newrelic.endTransaction()

### ISSUE THREAD

**Issue by infinity-naveen** [newrelic/node-newrelic#1775]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description
Main issues observed: higher `CPU utilisation`, higher max CPU time per tick
We were facing a major issue in our projects related to CPU utilisation after using `newrelic` nodejs agent in our app, we observed higher CPU utilisation and spikes in Max CPU time per tick graphs, whenever any long running transaction is completed. CPU was completely choked for few seconds just after completion of transaction handler. That transaction can be anything which is wrapped in newrelic transaction wrapper, so we faced issues almost everywhere, be it any repeatable job handlers, express API handlers, or other background executions. 

On further deep-diving into this issue, we found that `newrelic.endTransaction()` function is taking too much CPU in cases when transaction is having large no. of child segments (API calls/ DB calls/ `setTimeout`/ setInterval/ etc.) in internal code of transaction handler.

Example sample code to reproduce issue:
[`nr.endTransaction()` is called explicitly also, just to verify the issue]

```
const newrelic = require('newrelic')
const COUNT = process.env.COUNT || 25
function sampleTransactionHandle() {
  const tx = newrelic.getTransaction()
  const count = COUNT*1000; // complexity of endTransaction() depends on square proportional to this counter;
  for (let i =0; i<count; i++) {
    setTimeout(() => {}, 10); // just an example to create child segments of parent transaction.
  }
  console.time();
  tx.end(); // <---- CPU-intensive operation (Need to fix this) (operation time proportional to count^2)
  console.timeEnd();
  console.log('done')
  process.exit(0)
}
newrelic.startBackgroundTransaction('slowEnd', sampleTransactionHandle)
```

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT share sensitive information, whether personal, proprietary, or otherwise! )

## Expected Behavior
`endTransaction()` should not take much time (at max it should be linearly proportional to `count`)

[NOTE]: # ( Provide any other relevant log data. )
[TIP]:  # ( Scrub logs and diagnostic information for sensitive information )

## Steps to Reproduce
Execute above mentioned NodeJs code snippet.

Results:
On 3 sample runs:
count=25k  =>  time taken = 1.3 sec
count=50k  =>  time taken = 6.2 sec
count=100k => time taken = 25.2 sec

(time taken is square proportional to count)

## Your Environment 
Tried in multiple environments:
node: v12, v14, v18
newrelic: v7.5.2, v11.0.0 (latest)
OS: ubuntu, Mac

[TIP]:  # ( Include as many relevant details about your environment as possible. )

## Solution
On further debugging the issue, we were able to figure out issue in agent code, in which a function (used to calculate whole transaction time) is using Array.concat function internally which is increasing overall complexity. PR is raised to fix the same: https://github.com/newrelic/node-newrelic/pull/1774

## Additional context
There are some other npm packages like `util.promisify` or `safe-array-concat` - these mutate the array's concat function (using symbol `isConcatSpreadable`) & increase concat() complexity (but please note, complexity order remains same). Since, internally in newrelic agent code, concat() is used in a loop, it is causing more worser impact on CPU utilisation. In our case also, util.promisify was a sub-dependency in our project, which resulted in newrelic.endTransaction() more CPU-intensive.

[TIP]:  # ( Add any other context about the problem here. For example, relevant community posts or support tickets. )

**Comment by infinity-naveen** [comment]
Linking corresponding PR: https://github.com/newrelic/node-newrelic/pull/1774

**Comment by infinity-naveen** [comment]
closing this issue as it has been fixed in v11.1.0
